modules/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3RetryingInputStream.java

```diff
@@ -166,45 +177,110 @@ private void ensureOpen() {
     }
 
     private void reopenStreamOrFail(IOException e) throws IOException {
-        if (purpose == OperationPurpose.REPOSITORY_ANALYSIS) {
-            logger.warn(() -> format("""
-                failed reading [%s/%s] at offset [%s]""", blobStore.bucket(), blobKey, start + currentOffset), e);
-            throw e;
+        if (shouldRetry() == false) {
```

Comment:
nit You could potentially put this `if` inside `recordAndLogForRetry`, but I also do see the value of it being blatantly here for code readability.

Reply:
I moved this block into `recordAndLogForRetry` as well and renamed the method to `MaybeRecordAndLogForRetry`. Thanks!